software/contrib/probapoly.py

```diff
@@ -0,0 +1,242 @@
+
+button_1: Short press (<500ms): Reduce pattern length (using manual pattern length is ON). Long Press (>500ms): Toggle doubletime feature
+button_2: Short press (<500ms): Reduce pattern length (using manual pattern length is ON). Long Press (>500ms): Toggle Manual pattern length feature
+
+knob_1: Select upper polyrhythm value
+knob_2: Select lower polyrhythm value
+
+output_1: Gate upper polyrhythm
+output_2: Gate upper polyrhythm (50% probability)
+output_3: Gate upper polyrhythm (25% probability)
+output_4: Gate lower polyrhythm
+output_5: Gate lower polyrhythm (50% probability)
+output_6: Gate lower polyrhythm (50% probability)
+
+'''
+
+class Poly(EuroPiScript):
+    def __init__(self):
+        
+        # Needed if using europi_script
+        super().__init__()
+
+        # Overclock the Pico for improved performance.
+        machine.freq(250_000_000)
+
+        # Variables
+        self.step = 1
+        self.clockStep = 0
+        self.resetTimeout = 2000
+        self.maxPolyVal = 23
+        self.upper = 1
+        self.lower = 3
+        self.ainValue = 0
+        self.upperBernoulliProb = 50
+        self.lowerBernoulliProb = 50
+        self.upperProb1 = 50
+        self.upperProb2 = 25
+        self.lowerProb1 = 50
+        self.lowerProb2 = 25
+        self.doubleTime = False
+        self.doubleTimeManualOverride = False
+        self.manualPatternLength = False
+        self.patternLength = self.lcm(self.upper, self.lower)
+        self.patternLengthPrevious = self.patternLength
+        
+        # Todo: Make this mode accessible from the UI
+        # Mode 1: Analogue input toggles double time feature
+        # Mode 2: Analogue input voltage adjusts the upper poly value
+        # Mode 3: Analogue input voltage adjusts the probabilities of outputs 2,3,5,6 sending gates
+        self.ainMode = 3
+
+        @din.handler
+        def clockRising():
+            for cv in cvs:
+                cv.off()
+            #self.updateScreen()
+            self.handleClock()
+            self.clockStep +=1
+            self.step += 1
+
+            # Reached of of pattern, or a shorter patter is now needed, reset step to 0
+            if self.step >= self.patternLength + 1:
+                self.step = 1
+
+        @din.handler_falling
+        def clockFalling():
+            for cv in cvs:
+                cv.off()
+            if self.doubleTimeManualOverride or self.doubleTime:
+                #self.updateScreen()
+                self.handleClock()
+                self.clockStep +=1
+                self.step += 1
+
+                # Reached of of pattern, or a shorter patter is now needed, reset step to 0
+                if self.step >= self.patternLength + 1:
+                    self.step = 1
+
+        @b1.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b1.last_pressed()) >  500 and ticks_diff(ticks_ms(), b1.last_pressed()) < 3000:
```

Comment:
[optional] consider making `500` and `3000` constants to improve readability (the constant name adds self documenting code) and maintainability (you'll only need to change the value in one place to update both button handlers).

Reply:
Good shout. I will do this on my next commit.

Reply:
Done!